Add environment files for expanse

moxcodes opened the pull request:
## Proposed changes

Add environment file and submit script for running SpECTRE jobs on the expanse supercomputer.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

moxcodes committed:
Add environment files for expanse

markscheel commented on support/Environments/expanse_gcc.sh:
```diff
@@ -0,0 +1,374 @@
+    pip3 install --user --ignore-installed numpy
+    pip3 install --user scipy
+    pip3 install --user h5py
+
```
Do we want to install matplotlib?  One of the tests (Unit.Visualization.Python.Render1D) fails without it.  Not sure we care.

markscheel reviewed (commented):
This works for me on expanse.  3 of the unit tests there fail because of roundoff, and a 4th fails because of missing matplotlib.

markscheel commented:
Well I spoke too soon.  The unit tests all work on expanse (except as noted above).  But trying to run the EvolveGhKerrSchild executable on Expanse, with no changes to Expanse.sh (except for my SPECTRE_HOME) fails with the error.
`Charmrun> Timeout waiting for node-program to connect`

moxcodes commented:
hmm...oops I wonder how my setup ended up different -- I'll try to reproduce the failure.

markscheel commented:
And I get the same error if I try to run on one node instead of two.

moxcodes commented:
anything in `stderr`? I recall I get this error from charm if I've made a typo on the path to the executable sometimes

markscheel commented:
The entire spectre.stderr is:
```
Currently Loaded Modules:
  1) shared                  8) boost/1.74.0  15) libxsmm
  2) cpu/0.15.4              9) cmake/3.18.2  16) jemalloc
  3) DefaultModules         10) blaze         17) spectre_gsl
  4) slurm/expanse/20.02.3  11) brigand       18) yaml-cpp
  5) gcc/10.2.0             12) catch         19) scotch
  6) intel-mkl/2019.1.144   13) libsharp      20) charm
  7) gsl/2.5                14) hdf5

 

Charmrun> Timeout waiting for node-program to connect
```

markscheel commented:
spectre.stdout is also very short:
```
Running on the following nodes:
exp-2-29
Charmrun> scalable start enabled. 
Charmrun> IBVERBS version of charmrun
```

moxcodes committed:
Loosen approximations to pass tests on Expanse

moxcodes commented:
I've pushed a fixup and an extra commit to fix the roundoff-level errors and to include matplotlib.
I wasn't able to reproduce the failure to launch -- could you double-check that the `SPECTRE_BUILD_DIR` variable in the launch script appropriately points to where you built spectre? Note that unlike some of the submit scripts, this doesn't assume that the build directory is a subdirectory of `SPECTRE_HOME` -- I don't think that assumption is a good one for everyone's directory structure, but it's probably better for the scripts to be consistent, so once it's clear which version is preferable, I'll update the rest.

markscheel commented:
My build directory is not a subdirectory of SPECTRE_HOME.  I still find the same problem.

My SPECTRE_HOME and build directory and dependencies directory are in `/expanse/lustre/projects` instead of in my home dir.  Not sure why that would make a difference though.

moxcodes committed:
Add `SPECTRE_MODULE_DIR` variable to submit scripts

This allows the user to not put the `module use` statements in their
.bashrc, which helps support users managing multiple environemnts
on a given system

moxcodes commented:
Added an extra commit that adds a `SPECTRE_MODULE_DIR` variable to the submit scripts that use a `runscript` -- that's necessary to set up the environment on the compute nodes in a way that doesn't depend on the spectre modules being available via the users' `.bashrc`.
Cheers!

markscheel commented on support/Environments/expanse_gcc.sh:
```diff
@@ -0,0 +1,375 @@
+    printf "\n\nIMPORTANT!!!\nIn order to be able to use these modules you\n"
+    echo "must run:"
+    echo "  module use $dep_dir/modules"
+    echo "You will need to do this every time you compile SpECTRE, so you may"
+    echo "want to add it to your ~/.bashrc."
```
Do we still want to keep the message about adding this to your .bashrc?  Maybe, since you need it for compiling. But it might cause problems with people having more than one environment. (the way SpEC solves this is that its compile command loads the environment).

moxcodes replied:
I'll tweak the message to say "..., so if you want this SpECTRE environment to be your primary environment, you may want to add it to your ~/.bashrc."

markscheel reviewed (dismissed):
Looks good. Please squash.

moxcodes commented:
fixups and the alteration to the message about modules squashed in.
Cheers!
